mlflow/gateway/providers/openai.py

```diff
@@ -0,0 +1,185 @@
+from typing import Dict, Any
+
+from fastapi.encoders import jsonable_encoder
+
+from .base import BaseProvider
+from ..schemas import chat, completions, embeddings
+
+
+class OpenAIProvider(BaseProvider):
+    NAME = "openai"
+    SUPPORTED_ROUTES = ("chat", "completions", "embeddings")
+
+    async def _request(self, path: str, payload: Dict[str, Any]):
+        import aiohttp
+
+        config = self.config.model.config
+        token = config["openai_api_key"]
+        headers = {"Authorization": f"Bearer {token}"}
+        if org := config.get("openai_organization"):
+            headers["OpenAI-Organization"] = org
+        async with aiohttp.ClientSession(headers=headers) as session:
+            url = "/".join([config["openai_api_base"].rstrip("/"), path.lstrip("/")])
+            async with session.post(url, json=payload) as response:
+                response.raise_for_status()
+                return await response.json()
+
+    @staticmethod
+    def _make_payload(payload: Dict[str, Any], mapping: Dict[str, str]):
+        payload = payload.copy()
+        for k1, k2 in mapping.items():
+            if v := payload.pop(k1, None):
+                payload[k2] = v
+        return {k: v for k, v in payload.items() if v is not None and v != []}
+
+    async def chat(self, payload: chat.RequestPayload) -> chat.ResponsePayload:
+        payload = OpenAIProvider._make_payload(
+            jsonable_encoder(payload),
+            {"candidate_count": "n"},
+        )
+        resp = await self._request(
+            "chat/completions",
+            {"model": self.config.model.name, **payload},
+        )
+        # Response example (https://platform.openai.com/docs/api-reference/chat/create)
+        # ```
+        # {
+        #    "id":"chatcmpl-abc123",
+        #    "object":"chat.completion",
+        #    "created":1677858242,
+        #    "model":"gpt-3.5-turbo-0301",
+        #    "usage":{
+        #       "prompt_tokens":13,
+        #       "completion_tokens":7,
+        #       "total_tokens":20
+        #    },
+        #    "choices":[
+        #       {
+        #          "message":{
+        #             "role":"assistant",
+        #             "content":"\n\nThis is a test!"
+        #          },
+        #          "finish_reason":"stop",
+        #          "index":0
+        #       }
+        #    ]
+        # }
+        # ```
+        return chat.ResponsePayload(
+            **{
+                "candidates": [
+                    {
+                        "message": {
+                            "role": c["message"]["role"],
+                            "content": c["message"]["content"],
+                        },
+                        "metadata": {
+                            "finish_reason": c["finish_reason"],
```

Comment:
Looks good - I updated https://docs.google.com/document/d/1AEDlqSWh4n0j1Ue-llk03aRz7-xBK4P4Zh-Yvo1uhnI/edit to make the finish reason values either `stop` or `length` (consistent with OpenAI)

Reply:
Do other providers have different reasons?

Reply:
If necessary, I can update the response models. `finish_reason` is `Optional[str]`, but enum is more appropriate.

Reply:
Makes sense! We could use Enum to be more strict:

```diff
diff --git a/mlflow/gateway/schemas/chat.py b/mlflow/gateway/schemas/chat.py
index 8fff49774..42fe0c32c 100644
--- a/mlflow/gateway/schemas/chat.py
+++ b/mlflow/gateway/schemas/chat.py
@@ -1,3 +1,4 @@
+from enum import Enum
 from typing import List, Optional
 
 from pydantic import BaseModel, Extra, Field
@@ -21,8 +22,13 @@ class RequestPayload(BaseRequestPayload):
     messages: List[Message] = Field(..., min_items=1)
 
 
+class FinishReason(str, Enum):
+    STOP = "stop"
+    LENGTH = "length"
+
+
 class CandidateMetadata(BaseModel, extra=Extra.forbid):
-    finish_reason: Optional[str]
+    finish_reason: Optional[FinishReason]
 
 
 class Candidate(BaseModel):
diff --git a/mlflow/gateway/schemas/completions.py b/mlflow/gateway/schemas/completions.py
index 5034c80af..eca35357f 100644
--- a/mlflow/gateway/schemas/completions.py
+++ b/mlflow/gateway/schemas/completions.py
@@ -2,7 +2,7 @@ from typing import List, Optional
 
 from pydantic import BaseModel, Extra
 
-from .chat import BaseRequestPayload
+from .chat import BaseRequestPayload, FinishReason
 from ..config import RouteType
 
 
@@ -11,7 +11,7 @@ class RequestPayload(BaseRequestPayload):
 
 
 class CandidateMetadata(BaseModel, extra=Extra.forbid):
-    finish_reason: Optional[str]
+    finish_reason: Optional[FinishReason]
 
 
 class Candidate(BaseModel, extra=Extra.forbid):
```

<img width="884" alt="image" src="https://github.com/mlflow/mlflow/assets/17039389/ecff25db-4816-4fce-8375-abe63969c9e3">

You can see `stop` and `length` on the bottom right.

Reply:
Updated!